adr/2025-11-14-introduce-product-type-and-deprecate-states.md

```diff
@@ -0,0 +1,119 @@
+  - The rule conditions and product stream filters are tightly coupled to the legacy states (hard coded in both client-side and server-side), making it difficult to extend or modify their behavior. For e.g a third-party developer wanting to add a new product type, they would need to modify the existing rule conditions, product stream filters, product listing filters which is not ideal.
+
+## Decision
+
+### Deprecation of `product.states`:
+
+- Deprecate the `product.states` field in the database in favor of a new `product.type` field that clearly indicates whether a product is `digital` or `physical`.
+- Deprecate `order_line_item.states` in favor of `order_line_item.product_type` in a similar manner.
+- Deprecate `LineItemProductStatesRule` in favor of `LineItemProductTypeRule`.
+- Deprecate `StatesUpdater` service and its related dispatched events (`ProductStatesBeforeChangeEvent`, `ProductStatesChangedEvent`).
+- Deprecate product stream filters and product listing filters that rely on `product.states`, guiding users to use the new `product.type` field instead.
+
+### Introduce `product.type` field
+
+Product type field should have a clear definition: It represent the type of product, whether it's physical or digital or bundle or subscription etc, and it should be immutable once set.
+
+In a more detailed manner, we will make the following changes:
+
+- Add a dedicated `product.type` column (values `physical`/`digital`) with DAL exposure, new entity constants, defaulting to `physical`.
+- `product.type` is only written once after product save (@See `ProductTypeInitiator`) service. It means once a product is marked as `digital` or `physical`, it cannot be changed to another type, this also improves data integrity and not having to update the field on every save.
+- Also add `order_line_item.product_type` and populate it when line items are converted from the cart; `LineItemTransformer` also reconstructs legacy states when needed.
+- Introduce `LineItemProductTypeRule` for rule builder usage and deprecate the legacy `LineItemProductStatesRule`.
+- Rules automatically pick up custom product types registered via the shared registry (@See `ProductTypeRegistry`), so PHP-based conditions stay consistent with storefront/admin filters.
+
+#### Introduce a server-side `ProductTypeRegistry` 
+
+- This registry help both core rules and plugins can register additional product types via the `shopware.product.type` service tag.
+
+```php
+class ProductTypeRegistry
+{
+    /**
+     * @var array<string, AbstractProductType>
+     */
+    private array $types = [];
+
+    public function addType(AbstractProductType $type): void
+
+    public function getTypes(): array
+```
+
+The registry contains different `AbstractProductType` implementations, each defining:
+
+```php
+abstract class AbstractProductType
+{
+    abstract public function getType(): string;
+
+    public function getBehavior(): ProductTypeBehavior
```

Comment:
i don't see the benefit of providing the behaviour on the product type

Can you explain that in more details? E.g. what are the specific behaviours you can control etc

Providing that only makes sense if we somehow can provide a more or less complete behaviours, and we you add new types you can do that by mixing and matching of existing behaviours

however so far i think it is way more likely that when you add a new typ you would need to add new behaviours as well, which would then lead to a core change again 🤔